src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs

```diff
@@ -279,4 +290,99 @@ private SafeSslHandle() : base(IntPtr.Zero, true)
         {
         }
     }
+
+    internal sealed class SafeChannelBinding : SafeHandle
+    {
+        [StructLayout(LayoutKind.Sequential)]
+        private struct SEC_CHANNEL_BINDINGS
+        {
+            internal int dwInitiatorAddrType;
+            internal int cbInitiatorLength;
+            internal int dwInitiatorOffset;
+            internal int dwAcceptorAddrType;
+            internal int cbAcceptorLength;
+            internal int dwAcceptorOffset;
+            internal int cbApplicationDataLength;
+            internal int dwApplicationDataOffset;
+        }
+
+        private static readonly byte[] s_tlsServerEndPointByteArray = Encoding.UTF8.GetBytes("tls-server-end-point:");
+        private static readonly byte[] s_tlsUniqueByteArray = Encoding.UTF8.GetBytes("tls-unique:");
+        private static readonly int s_secChannelBindingSize = Marshal.SizeOf<SEC_CHANNEL_BINDINGS>();
+        private readonly int _cbtPrefixByteArraySize;
+        private const int CertHashMaxSize = 128;
+
+        internal int Length
+        {
+            get;
+            private set;
+        }
+
+        internal IntPtr CertHashPtr
+        {
+            get;
+            private set;
+        }
+
+        internal void SetCertHash(byte[] certHashBytes)
+        {
+            Debug.Assert(certHashBytes != null, "check certHashBytes is not null");
+            int length = certHashBytes.Length;
+            Marshal.Copy(certHashBytes, 0, CertHashPtr, length);
+            SetCertHashLength(length);
+        }
+
+        private byte[] GetPrefixBytes(ChannelBindingKind kind)
+        {
+            if (kind == ChannelBindingKind.Endpoint)
+            {
+                return s_tlsServerEndPointByteArray;
+            }
+            else if (kind == ChannelBindingKind.Unique)
+            {
+                return s_tlsUniqueByteArray;
+            }
+            else
+            {
+                throw new NotSupportedException();
+            }
+        }
+
+        internal SafeChannelBinding(ChannelBindingKind kind)
+            : base(IntPtr.Zero, true)
+        {
+            byte[] cbtPrefix = GetPrefixBytes(kind);
+            _cbtPrefixByteArraySize = cbtPrefix.Length;
+            handle = Marshal.AllocHGlobal(s_secChannelBindingSize + _cbtPrefixByteArraySize + CertHashMaxSize);
+            IntPtr cbtPrefixPtr = handle + s_secChannelBindingSize;
+            Marshal.Copy(cbtPrefix, 0, cbtPrefixPtr, _cbtPrefixByteArraySize);
+            CertHashPtr = cbtPrefixPtr + _cbtPrefixByteArraySize;
```

Comment:
Have few doubts related to this.
1- we can create the byte[] with new byte[s_secChannelBindingSize + _cbtPrefixByteArraySize + CertHashMaxSize] instead of Marshal.AllocHGlobal. But while calling SslGetPeerFinished & SslGetFinished, we will have to do
unsafe
{
  fixed(byte* ptr = byteArray)
  {
       //this will be the correct offset to start the certHash.
       var finalPtr = ptr + s_secChannelBindingSize + _cbtPrefixByteArraySize
      SslGetPeerFinished(finalPtr) or SslGetFinished(finalPtr)
  }
}

Please confirm.

2. Ultimately this will be used by enduser in the form of handle only. like
  
ChannelBinding obj = stream.GetChannelBinding(ChannelBindingKind.Endpoint);
obj.DangerousAddRef(ref bool);
obj.DangerousGetHandle();

 So we will have to keep the byte[] in the SafeChannelBindingHandle object & assign it's pointer to the SafeFreeContextBufferChannelBinding handle. Also this would mean that DangerousAddRef() would be sort of redundant

Looks like even if we allocate memory in the managed, we will end up doing back & forth with Native representation. That is the reason, I am not very convinced with allocating managed. But if we are ok with above points, I can change this to managed.